File: apps/shared/src/commonMain/kotlin/dev/supermux/ui/ChatDetail.kt
package dev.supermux.ui

/**
 * Chat transcript density levels (web/iOS/Android/Desktop/macOS parity).
 * High: terminal windows + file diffs for Bash/Edit/Write; quiet rows for other tools.
 */
enum class ChatDetailLevel {
    LOW,
    MEDIUM,
    HIGH,
    ;

    val wire: String
        get() = when (this) {
            LOW -> "low"
            MEDIUM -> "medium"
            HIGH -> "high"
        }

    val label: String
        get() = when (this) {
            LOW -> "Low"
            MEDIUM -> "Medium"
            HIGH -> "High"
        }

    companion object {
        fun parse(raw: String?): ChatDetailLevel = when (raw?.lowercase()) {
            "low" -> LOW
            "medium" -> MEDIUM
            "high" -> HIGH
            else -> MEDIUM
        }
    }
}

/** All three levels are implemented. */
fun isChatDetailImplemented(level: ChatDetailLevel): Boolean =
    level == ChatDetailLevel.LOW || level == ChatDetailLevel.MEDIUM || level == ChatDetailLevel.HIGH

/** Effective render mode — all implemented levels pass through. */
fun effectiveChatDetail(level: ChatDetailLevel): ChatDetailLevel =
    when (level) {
        ChatDetailLevel.LOW -> ChatDetailLevel.LOW
        ChatDetailLevel.HIGH -> ChatDetailLevel.HIGH
        else -> ChatDetailLevel.MEDIUM
    }

/**
 * Reject unimplemented levels for set-level APIs.
 * @return the level to store, or null if the request should no-op.
 */
fun sanitizeSetLevel(level: ChatDetailLevel): ChatDetailLevel? =
    if (isChatDetailImplemented(level)) level else null

/**
 * Turn boundary for "tools this turn": last **user** message timestamp (ms),
 * else [workingSinceMs], else 0.
 *
 * Native clients treat user messages as `direction == "inbound"` (or `hasPrefix("in")`).
 * Pass [isUserDirection] accordingly. [tsToEpochMs] converts wire `ts` to epoch ms.
 */
fun turnBoundaryMs(
    messages: List<Pair<String, String>>, // direction, ts
    isUserDirection: (String) -> Boolean,
    tsToEpochMs: (String) -> Long,
    workingSinceMs: Long? = null,
): Long {
    for (i in messages.indices.reversed()) {
        val (dir, ts) = messages[i]
        if (isUserDirection(dir)) return tsToEpochMs(ts)
    }
    return workingSinceMs ?: 0L
}

/** Count tool events at or after [sinceMs]. */
fun countToolsSince(
    toolTimestampsMs: List<Long>,
    sinceMs: Long,
): Int = toolTimestampsMs.count { it >= sinceMs }

/**
 * Low-mode working status: platform [baseLabel] + optional tool + count + duration.
 * Segments joined with " · ".
 */
fun formatLowWorkingStatus(
    baseLabel: String,
    detail: String?,
    tool: String?,
    toolCount: Int,
    durationLabel: String,
): String {
    val parts = mutableListOf(baseLabel)
    if (detail == "running" && !tool.isNullOrBlank()) parts.add(tool)
    if (toolCount > 0) {
        parts.add(if (toolCount == 1) "1 tool" else "$toolCount tools")
    }
    if (durationLabel.isNotEmpty()) parts.add(durationLabel)
    return parts.joinToString(" · ")
}

File: src/web-app/src/lib/chat-detail.ts
/** Chat transcript density: low | medium | high. */

export type ChatDetailLevel = "low" | "medium" | "high"

export const CHAT_DETAIL_LEVELS = ["low", "medium", "high"] as const satisfies readonly ChatDetailLevel[]

export const CHAT_DETAIL_DEFAULT: ChatDetailLevel = "medium"

export const CHAT_DETAIL_LABELS: Record<ChatDetailLevel, string> = {
  low: "Low",
  medium: "Medium",
  high: "High",
}

/** Levels that are fully implemented for rendering. */
export function isChatDetailImplemented(level: ChatDetailLevel): boolean {
  return level === "low" || level === "medium" || level === "high"
}

/** Effective render mode — all three levels are live. */
export function effectiveChatDetail(level: ChatDetailLevel): ChatDetailLevel {
  if (level === "low" || level === "high") return level
  return "medium"
}

/** Persistable levels — garbage clamps to medium; high is accepted. */
export function parseChatDetailLevel(raw: unknown): ChatDetailLevel {
  if (raw === "low" || raw === "medium" || raw === "high") return raw
  return "medium"
}

/** Turn boundary for "tools this turn": last outbound user message ts, else workingSince, else 0. */
export function turnBoundaryMs(
  messages: { direction: string; ts: string }[],
  workingSince?: number | null,
): number {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i]!.direction === "outbound") {
      return new Date(messages[i]!.ts).getTime()
    }
  }
  return workingSince ?? 0
}

export function countToolsSince(
  acts: { kind: string; ts: string }[],
  sinceMs: number,
): number {
  return acts.filter((e) => e.kind === "tool" && new Date(e.ts).getTime() >= sinceMs).length
}

export type LowWorkingStatusInput = {
  /** Platform baseline, e.g. "Working…" */
  baseLabel: string
  detail?: "thinking" | "running" | null
  tool?: string | null
  toolCount: number
  /** Already formatted, e.g. formatDuration output */
  durationLabel: string
}

/**
 * Segments after baseLabel, joined with " · ".
 * - No tools segment when toolCount === 0
 * - Tool name only when detail === "running" && tool
 * - Duration last when non-empty
 */
export function formatLowWorkingStatus(input: LowWorkingStatusInput): string {
  const parts: string[] = [input.baseLabel]
  if (input.detail === "running" && input.tool) parts.push(input.tool)
  if (input.toolCount > 0) {
    parts.push(input.toolCount === 1 ? "1 tool" : `${input.toolCount} tools`)
  }
  if (input.durationLabel) parts.push(input.durationLabel)
  return parts.join(" · ")
}
